lib/Runtime/Executor/ThreadPoolExecutor.cpp

```diff
@@ -80,13 +80,38 @@ ExecutionState::ExecutionState(RunIdentifierTy id, const DAGNode *root,
     nodeParentsDone_[node] = 0;
 
     // Make an (empty) input context for the node.
-    auto ctx = llvm::make_unique<ExecutionContext>();
+    auto nodeInputCtx = llvm::make_unique<ExecutionContext>();
+
     if (resultTraceContext) {
-      ctx->setTraceContext(llvm::make_unique<TraceContext>(
+      nodeInputCtx->setTraceContext(llvm::make_unique<TraceContext>(
           resultTraceContext->getTraceLevel(),
           resultTraceContext->getTraceThread()));
     }
-    inputCtxs_.insert(std::make_pair(node, std::move(ctx)));
+
+    auto nodeInputPhBindings = nodeInputCtx->getPlaceholderBindings();
+
+    // Get the symbol table for the node.
+    const SymbolTableTy &symbolTable = node->runtimeBundle->getSymbolTable();
+
+    // Create Placeholders for the outputs of all intermediate nodes. These are
+    // not in the ExecutionContext passed to Executor::run, so they must be
+    // created by the Executor. Only output symbols need to be considered
+    // because the Placeholders created for them will be copied to the input
+    // PlaceholderBindings of child nodes by
+    // ThreadPoolExecutor::propagatePlaceholdersForNode.
+    for (const auto &symbolPair : symbolTable) {
+      const auto &symbolName = symbolPair.first;
+      const auto &symbolInfo = symbolPair.second;
+
+      if (symbolInfo.output) {
```

Comment:
So we create a context and add it's output PH's here, then we insert the input placeholders in `propagatePlaceholdersForNode`. Would it make more sense to build the entire context at once? Then just focus on copying data in/out?